ethereum/datastructures/src/main/java/tech/pegasys/teku/datastructures/blocks/BeaconBlockHeader.java

```diff
@@ -87,6 +94,43 @@ public BeaconBlockHeader() {
     super(TYPE);
   }
 
+  /**
+   * Returns the block header associated with this state
+   *
+   * @param state A beacon state
+   * @return The latest block header from the state, with stateRoot pointing to the supplied state
+   */
+  public static BeaconBlockHeader fromState(final BeaconState state) {
+    BeaconBlockHeader latestHeader = state.getLatest_block_header();
+
+    if (latestHeader.getStateRoot().isZero()) {
+      // If the state root is empty, replace it with the current state root
+      final Bytes32 stateRoot = state.hash_tree_root();
+      latestHeader =
+          new BeaconBlockHeader(
+              latestHeader.getSlot(),
+              latestHeader.getProposerIndex(),
+              latestHeader.getParentRoot(),
+              stateRoot,
+              latestHeader.getBodyRoot());
+    }
+
+    return latestHeader;
+  }
+
+  public static BeaconBlockHeader fromBlock(final BeaconBlock block) {
+    return new BeaconBlockHeader(
+        block.getSlot(),
+        block.getProposerIndex(),
+        block.getParentRoot(),
+        block.getStateRoot(),
+        block.getBody().hash_tree_root());
```

Comment:
nit: I think you can just use:
```suggestion
        block.getBodyRoot());
```